### DISCUSSION TITLE
MPI, threads and Minimum Clang

### DISCUSSION THREAD

**Discussion by WilkAndy** [idaholab/moose#16098] [Q&A General] [answered]
Hi everyone,

I'm hoping someone might have thoughts on how to fix a bug.  A recent PR of mine, #15698 , is not passing the "Minimum Clang Version".  All other recipes pass.  The problem is in a test that has

```
    min_threads = 2
    min_parallel = 2
 ```

The same test passes on Minimum Clang with just `min_threads = 2` and with just `min_parallel = 2`, but not both threads and MPI active.  Also, this test has 2 timesteps, each producing a file from a `LineValueSampler`.  The file from the first timestep passes, but the single node that is on the postprocessor boundary yields the incorrect result on the second timestep.   I'm not 100% sure, but i believe that the cause is that NodalUserObject::execute() is not being run for this node at the second timestep.

So, two questions:

1. any thoughts about what's going wrong?  I know this is difficult to answer without diving into my code, which you do NOT want to do!

2. any thoughts about how to debug this?  I can't reproduce the error on any of my systems, and obviously CIVET can only produce the error on Minimum Clang.

**Comment by lindsayad** [comment]
Does this happen every time on CIVET or only sometimes? It would be a blessing if it’s every time; then the issue is less likely to be a dreaded race condition. I typically debug race conditions with helgrind but that’s kind of difficult to get setup...you have to ensure your threads are entirely pthreads based

**Comment by WilkAndy** [comment]
I think it occurs every time.  What do you mean about pthreads?  I thought we couldn't control that sort of thing with MOOSE?  I just use NodalUserObject::threadJoin without thinking about behind-the-scenes stuff.

**Reply by lindsayad** [reply]
What thread library is used behind the scenes is controlled with libmesh configure options

**Reply by WilkAndy** [reply]
Oh, OK.  Just to eliminate this possibility: do you know whether the Minimum Clang recipe uses a different thread library to the other recipes?   My uneducated opinion is that the problem is unlikely to be the thread library since the test only fails when both MPI and threads are active, but passes otherwise (only MPI, and only threads, and neither)

**Reply by permcody** [reply]
Minimum Clang definitely did use a different thread implementation because of missing support, not because of an explicit selection of configuration flags. This may not be the case anymore as we've recently bumped the minimum version of things. Can you give us a little more information on what's going on in that test or which post process appears to be failing? I wouldn't expect any of the thread libraries to cause a specific failure like this if your postprocessor is implemented correctly.

**Reply by lindsayad** [reply]
GitHub discussions don't appear to show mentions from issues/PRs :frowning: Now we can have trifurcated conversations. Conversation also over here: https://github.com/idaholab/moose/pull/15698#issuecomment-724161778 We do explicitly request `--with-thread-model=pthread` in that recipe, but we don't explicitly `--disable-openmp`. However, @permcody is right: that box doesn't appear to support openmp, so if you look at the configuration summary, there is no `-fopenmp` in `CXXFLAGS`. On my desktop, if I pass `--with-thread-model=pthread`, I still get `-fopenmp`.

**Reply by lindsayad** [reply]
Mmm also mentions of issues/PRs from discussions doesn't trigger an event on the referenced issue/PR either. I don't like that

**Reply by permcody** [reply]
Another relevant issue: https://github.com/libMesh/libmesh/issues/2070

**Comment by WilkAndy** [comment]
Another note to this Discussion.  We have discovered that the test fails irrespective of whether MPI is used or not - ie spatial_4_threads and spatial_4_threads_MPI BOTH fail.  It probably makes debugging easier to know that MPI is not involved, and it could point to a race condition as the spatial_4_threads always (luckily) passed until now.

They are being skipped at #16122

**Comment by WilkAndy** [comment]
@lindsayad , thanks for your offer of help.  I want to have a look a bit first.  But could you please tell me why many modules tests have threads != pthreads ?  I've seen that so many times, and can't remember why it's there.  Perhaps pthreads are known to be crappy with MOOSE?

**Reply by lindsayad** [reply]
Hey look at this guy: https://github.com/idaholab/moose/pull/11977#issuecomment-412377652. You heard of him? :wink: So from [later on in that thread](https://github.com/idaholab/moose/pull/11977#issuecomment-412556778), you have an explanation from @andrsd. ~I myself would like to know quantitatively what "tonnes of threads" means.~ Also we should definitely be erroring out if we are attempting to run pthreads on code that is not safe for pthreads.

**Reply by WilkAndy** [reply]
Thanks for the info about pthreads and GeneralUserObjects.

This current problem happens for a NodalUserObject, so i believe it's not relevant.

**Reply by lindsayad** [reply]
Correct, not relevant to your thread problem. But that is why there are so many instances of `threading != pthreads` in modules

**Comment by WilkAndy** [comment]
Hey everyone, just another small question - i've just opened #16128 to attempt to explore this problem.  I actually only want to run "Minimum Clang" and no other recipe.  Is that somehow possible?

**Reply by permcody** [reply]
Short answer is that we don't have a direct way but there is a solution that at least administrators can exercise:
1. Cancel all jobs
2. Invalidate just the one job you are interested in seeing.

**Reply by permcody** [reply]
Seems like a good feature request. Hopefully my new technician (starting next month) may have time to start looking at expanding capabilities a bit here and there.

**Comment by WilkAndy** [comment, marked as answer]
I think i've got to the bottom of this, but don't have an adequate fix.  Thanks for your help @lindsayad and @permcody - perhaps you have suggestions on how to modify my code, or the tests or the framework.

I believe it occurs because the threads do not stay attached to nodes between timesteps.  Let me try to explain.

I have a NodalUserObject that computes a `GeochemistrySpatialReactor egs_at_node` at each node.  `egs_at_node` is different at each node and each timestep.  At each timestep, i copy the values of `egs_at_node` from the "child threads" to the "main thread" using `threadJoin`, at https://github.com/idaholab/moose/blob/30329172406a5dd9179b37dba3b979b0ee7b23e9/modules/geochemistry/src/userobjects/GeochemistrySpatialReactor.C#L422-L428 .   ("child threads" and "main thread" are my names - i hope i've interpreted `threadJoin` correctly.)

During initialization of the simulation, `egs_at_node` gets initialized in all threads.  Now consider the case where a main thread looks after node=0 for the first N timesteps.  Then `threadJoin` doesn't do anything for `egs_at_node[0]` (ie, at node=0) at these timesteps, which is as desired.  Now suppose that a child thread looks after node=0 at timestep N+1.  The problem is that its copy of `egs_at_node[0]` is still in the initialized state: it hasn't ever been touched by `execute()`.  So `threadJoin` copies this `egs_at_node[0]` to the main thread, which overwrites it with incorrect values.

Hopefully this makes sense.

I don't know for 100% sure that this is happening, but my mucking around at line 436 of https://github.com/idaholab/moose/blob/237169900b6a94d7090e186d4cd97a0f04e101c7/modules/geochemistry/src/userobjects/GeochemistrySpatialReactor.C#L422-L443 suggests that it's true (the mucking around only works for 2 threads).

So, where to go from here?

**Reply by permcody** [reply]
Correct Andy. Threads are not guaranteed to have any sort of affinity so this is indeed likely the case. You should treat threads as independent processing tasks that are somewhat disconnected to the underlying memory layout. In the general case the OS can select any thread that's available and assign it to a task. As you've noticed this isn't always the case. With some models we do have thread affinity, but other we do not. 

Option 1: You can create a map instead of an ordered vector to handle your contributions back into your main data structure. That way no matter which thread handles a particular request, you'll be assured it ends up in the right spot during the join.

Option 2: You can create a global data structure (static) and sum directly into it. Note that you do have to be careful about this kind of approach. Is your data structure thread safe meaning it can handle concurrent writes, or do you have to implement some sort of locking mechanism. In MOOSE we have a little of both. Depending on your needs we do have a thread-safe "hash map" available and there are also lock abstractions you can use to write to the master data structure.

**Reply by WilkAndy** [reply]
Thanks for those options, @permcody.   I don't really understand how each of them work.  Sorry for being so ignorant.  At the moment i have things like https://github.com/idaholab/moose/blob/b908c63e889276225c2fbc26a91888b2c2ac146e/modules/geochemistry/include/userobjects/GeochemistrySpatialReactor.h#L51-L52  

Option 1: Are you saying this should somehow be a std::map with keys being node numbers and values being `GeochemicalSystem` ?   I don't understand how this is going to help.  Won't all the threads have their own copy of this map?

Option 2:  Are you saying i should just write `static` in front of this line?    I think std::vectors are thread safe.   Then I don't have to do anything in threadJoin?

**Reply by WilkAndy** [reply]
Don't forget to reply at some stage :-)

**Reply by GiudGiud** [reply]
He forgot ^^. @permcody 
static only provides thread safety at initialization. You'd still need a lock to modify it. If threads are just reading from it, then no worries. 
It's also quite common in moose to have everything duplicated for each threads, to avoid locks. 

@WilkAndy can this be closed?

**Comment by WilkAndy** [comment]
Thanks @GiudGiud , this may be closed because of #16133 .   In summary: main thread gets info from the other threads during GeochemistrySpatialReactor::threadJoin, so main thread has all correct info after threadJoin.  Then, by #16133 , during the main-thread call of GeochemistrySpatialReactor::finalize(), the other-thread info is set equal to the main-thread info.  Hence, after finalize(), all threads have correct info.
